### PR TITLE
Generator with options

### PR DESCRIPTION
**Pull request by LeonanCarvalho**
I rewrote the PR https://github.com/gofrs/uuid/pull/98 authored by @mlesar  based on the @theckman  comments and recommendations.

Help is wanted to make it happens, this is a very desirable feature especially to generate UUIDv6 with custom timestamps.

### CODE REVIEW

**Review by baldanca — APPROVED** [review]

File: generator.go [hunk 1]
```diff
@@ -147,11 +151,57 @@ func NewGen() *Gen {
...
+		opt(gen)
+	}
+
+	return gen
+}
+
+// WithHWAddrFunc is a GenOption that allows you to provide your own HWAddrFunc function
+func WithHWAddrFunc(hwaf HWAddrFunc) GenOption {
+	return func(gen *Gen) {
+		gen.hwAddrFunc = hwaf
```
Comment (dylan-bourque) [inline generator.go:189]: This needs a nil check.  `WithHWAddrFunc(nil)` will compile but lead to an immediate panic when the generator is used.

File: generator.go [hunk 2]
```diff
@@ -147,11 +151,57 @@ func NewGen() *Gen {
...
+	}
+}
+
+// WithEpochFunc is a GenOption that allows you to provide your own EpochFunc function
+func WithEpochFunc(epochf EpochFunc) GenOption {
+	return func(gen *Gen) {
+		gen.epochFunc = epochf
```
Comment (dylan-bourque) [inline generator.go:196]: Same here, need to check if `epochf` is `nil`.

File: generator.go [hunk 3]
```diff
@@ -147,11 +151,57 @@ func NewGen() *Gen {
...
+	}
+}
+
+// WithRandomReader is a GenOption that allows you to provide your own random reader
+func WithRandomReader(reader io.Reader) GenOption {
+	return func(gen *Gen) {
+		gen.rand = reader
```
Comment (dylan-bourque) [inline generator.go:203]: And here.

**Review by dylan-bourque — CHANGES_REQUESTED** [review]

File: generator.go (hunk 1 above)
Comment (LeonanCarvalho) [reply generator.go:189]: We do not need to validate each option, instead, we can bypass `NewGenWithOptions` when the `nil` option has been passed. 

```go
func NewGenWithOptions(opts ...GenOption) *Gen {
	gen := &Gen{
		epochFunc:  time.Now,
		hwAddrFunc: defaultHWAddrFunc,
		rand:       rand.Reader,
	}

	for _, opt := range opts {
		if opt != nil {
			opt(gen)
		}
	}

	return gen
}
```

What do you think?

Comment (dylan-bourque) [reply generator.go:191]: In the example I gave the option itself won't be `nil` but it would end up setting the `Gen.hwAddrFunc` field to `nil`.  The `if opt != nil` check will never be false unless someone explicitly passes `GenOption(nil)`.

To be 100% safe we should be checking the option itself in `NewGenWithOptions()` and checking for `nil` **values** in the option functions themselves.

```
func WithHWAddrFunc(hwaf HWAddrFunc) GenOption {
    return func(gen *Gen) {
        if hwaf != nil {
            gen.hwAddrFunc = hwaf
        }
    }
}
```
or
```
func WithHWAddrFunc(hwaf HWAddrFunc) GenOption {
    return func(gen *Gen) {
        if hwaf == nil {
            hwaf = defaultHWAddrFunc
        }
        gen.hwAddrFunc = hwaf
    }
}
```

Comment (LeonanCarvalho) [reply generator.go:193]: Got it!

Changed

File: generator.go [hunk 4]
```diff
@@ -147,11 +151,69 @@ func NewGen() *Gen {
...
 	}
+
+	for _, opt := range opts {
+		opt(gen)
+	}
+
+	return gen
+}
+
+// WithHWAddrFunc is a GenOption that allows you to provide your own HWAddrFunc function
```
Comment (dylan-bourque) [inline generator.go:186]: I would document the behavior when/if `hwaf` is `nil`.  Something along the lines of "If hwaf is nil a default function is used that returns the host's MAC address."

Similar for the two other options.

Comment (LeonanCarvalho) [reply generator.go]: Good point, fixed

**Review by dylan-bourque — APPROVED** [review]
LGTM

**Comment by LeonanCarvalho** [comment]
@theckman @cameracker  It would be great to have your review too :)

File: generator_test.go [hunk 5]
```diff
@@ -175,6 +190,33 @@ func testNewV1MissingNetworkFaultyRand(t *testing.T) {
...
+	g := NewGenWithOptions(
+		WithHWAddrFunc(func() (net.HardwareAddr, error) {
+			return []byte{}, fmt.Errorf("uuid: no hw address found")
+		}),
+		WithRandomReader(&faultyReader{
+			readToFail: 1,
+		}),
+	)
+
+	u, err := g.NewV1()
```
Comment (cameracker) [inline generator_test.go:203]: any value in checking the error message?

**Comment by LeonanCarvalho** [comment]
> @LeonanCarvalho thanks for the submission. What are your thoughts on the decrease of code coverage? Can we get back to 100%?

I think we could include tests for nil values since there is an if with default's fallback suggested here https://github.com/gofrs/uuid/pull/111#discussion_r1057721608


https://app.codecov.io/gh/gofrs/uuid/pull/111

File: generator_test.go (hunk 5 above)
Comment (LeonanCarvalho) [reply generator_test.go:203]: I can't imagine what else to check at this point. I just kept the same behaviour as the existing tests, I didn't want to modify anything pre-existing.

**Comment by LeonanCarvalho** [comment]
Does something flaky on those tests? I ran it on my repo and it passed 
https://github.com/LeonanCarvalho/uuid/actions/runs/3788789705

**Comment by LeonanCarvalho** [comment]
I found out that pkg x/tools cover is no longer active and marked as deprecated, that's why the test is falling https://pkg.go.dev/golang.org/x/tools/cmd/cover

**Comment by LeonanCarvalho** [comment]
I opened another PR to solve this workflow issue https://github.com/gofrs/uuid/pull/115
